File: Boare.Lib.Vsq/LyricHandle.cs
/*
 * LyricHandle.cs
 * Copyright (c) 2008-2009 kbinani
 *
 * This file is part of Boare.Lib.Vsq.
 *
 * Boare.Lib.Vsq is free software; you can redistribute it and/or
 * modify it under the terms of the BSD License.
 *
 * Boare.Lib.Vsq is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 */
#if JAVA
package org.kbinani.vsq;

import java.io.*;
import java.util.*;
#else
using System;
using bocoree.java.util;

namespace Boare.Lib.Vsq {
#endif

#if JAVA
    public class LyricHandle implements Cloneable, Serializable{
#else
    [Serializable]
    public class LyricHandle : ICloneable {
#endif
        public Lyric L0;
        public int Index;
        public Vector<Lyric> Trailing = new Vector<Lyric>();

        public LyricHandle() {
            L0 = new Lyric();
        }

        public Lyric getLyricAt( int index ){
            if( index == 0 ){
                return L0;
            }else{
                return Trailing.get( index - 1 );
            }
        }

        public void setLyricAt( int index, Lyric value ){
            if( index == 0 ){
                L0 = value;
            }else{
                Trailing.set( index - 1, value );
            }
        }

        public int getCount(){
            return Trailing.size() + 1;
        }

        /// <summary>
        /// type = Lyric用のhandleのコンストラクタ
        /// </summary>
        /// <param name="phrase">歌詞</param>
        /// <param name="phonetic_symbol">発音記号</param>
        public LyricHandle( String phrase, String phonetic_symbol ) {
            L0 = new Lyric( phrase, phonetic_symbol );
        }

        public Object clone() {
            LyricHandle ret = new LyricHandle();
            ret.Index = Index;
            ret.L0 = (Lyric)L0.clone();
            int c = Trailing.size();
            for( int i = 0; i < c; i++ ){
                Lyric buf = (Lyric)Trailing.get( i ).clone();
                ret.Trailing.add( buf );
            }
            return ret;
        }

#if !JAVA
        public object Clone() {
            return clone();
        }
#endif

        public VsqHandle castToVsqHandle() {
            VsqHandle ret = new VsqHandle();
            ret.m_type = VsqHandleType.Lyric;
            ret.L0 = (Lyric)L0.clone();
            ret.Trailing = Trailing;
            ret.Index = Index;
            return ret;
        }
    }

#if !JAVA
}
#endif

File: port_js/org/kbinani/vsq/LyricHandle.js
/*
 * LyricHandle.js
 * Copyright (C) 2008-2010 kbinani
 *
 * This file is part of org.kbinani.vsq.
 *
 * org.kbinani.vsq is free software; you can redistribute it and/or
 * modify it under the terms of the GPLv3 License.
 *
 * org.kbinani.vsq is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 */
if( org == undefined ) var org = {};
if( org.kbinani == undefined ) org.kbinani = {};
if( org.kbinani.vsq == undefined ) org.kbinani.vsq = {};
if( org.kbinani.vsq.LyricHandle == undefined ){

    /// <summary>
    /// type = Lyric用のhandleのコンストラクタ
    /// </summary>
    /// <param name="phrase">歌詞</param>
    /// <param name="phonetic_symbol">発音記号</param>
    org.kbinani.vsq.LyricHandle = function(){
        if( arguments.length >= 2 ){
            var phrase = arguments[0];
            var phonetic_symbol = arguments[1];
            this.L0 = new org.kbinani.vsq.Lyric( phrase, phonetic_symbol );
        }else{
            this.L0 = new org.kbinani.vsq.Lyric();
        }
        this.Index = 0;
        this.Trailing = new Array();
    };

    org.kbinani.vsq.LyricHandle.prototype = {
        /**
         * @param index [int]
         * @return [Lyric]
         */
        getLyricAt : function( index ){
            if( index == 0 ){
                return this.L0;
            }else{
                return this.Trailing[index - 1];
            }
        },

        /**
         * @param index [int]
         * @param value [Lyric]
         * @return [void]
         */
        setLyricAt : function( index, value ){
            if( index == 0 ){
                this.L0 = value;
            }else{
                this.Trailing[index - 1] = value;
            }
        },

        /**
         * @return [int]
         */
        getCount : function(){
            return this.Trailing.length + 1;
        },

        /**
         * @return [Lyric]
         */
        clone : function() {
            var ret = new org.kbinani.vsq.LyricHandle();
            ret.Index = this.Index;
            ret.L0 = this.L0.clone();
            var c = this.Trailing.length;
            for( var i = 0; i < c; i++ ){
                var buf = this.Trailing[i].clone();
                ret.Trailing.push( buf );
            }
            return ret;
        },

        /**
         * @return [VsqHandle]
         */
        castToVsqHandle : function() {
            var ret = new VsqHandle();
            ret.m_type = org.kbinani.vsq.VsqHandleType.Lyric;
            ret.L0 = this.L0.clone();
            ret.Trailing = this.Trailing;
            ret.Index = this.Index;
            return ret;
        },
    };

}
